subprojects/core/src/main/java/org/gradle/api/internal/changedetection/changes/DefaultTaskExecutionMode.java

```diff
@@ -0,0 +1,129 @@
+    private final boolean taskHistoryMaintained;
+    private final boolean allowedToUseCachedResults;
+
+    DefaultTaskExecutionMode(@Nullable String rebuildReason, boolean taskHistoryMaintained, boolean allowedToUseCachedResults) {
+        this.rebuildReason = Optional.ofNullable(rebuildReason);
+        this.taskHistoryMaintained = taskHistoryMaintained;
+        this.allowedToUseCachedResults = allowedToUseCachedResults;
+    }
+
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Optional<String> getRebuildReason() {
+        return rebuildReason;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isTaskHistoryMaintained() {
+        return taskHistoryMaintained;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isAllowedToUseCachedResults() {
+        return allowedToUseCachedResults;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        DefaultTaskExecutionMode that = (DefaultTaskExecutionMode) o;
+        return taskHistoryMaintained == that.taskHistoryMaintained && allowedToUseCachedResults == that.allowedToUseCachedResults && Objects.equals(rebuildReason, that.rebuildReason);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(rebuildReason, taskHistoryMaintained, allowedToUseCachedResults);
+    }
+
+    /**
+     * The execution mode for incremental tasks.
+     */
+    public static TaskExecutionMode incremental() {
+        return new DefaultTaskExecutionMode(null, true, true);
+    }
+
+    /**
+     * The execution mode when task did not declare any outputs.
+     * The message will be `Task has not declared any outputs despite executing actions.`.
+     */
+    public static TaskExecutionMode noOutputs() {
+        return new DefaultTaskExecutionMode("Task has not declared any outputs despite executing actions.", false, false);
+    }
+
+    /**
+     * The execution mode when the command was run with --rerun-tasks.
+     * The message will be `Executed with '--rerun-tasks'.`.
+     */
+    public static TaskExecutionMode rerunTasksEnabled() {
+        return new DefaultTaskExecutionMode("Executed with '--rerun-tasks'.", true, false);
+    }
+
+    /**
+     * The execution mode when the Task.upToDateWhen is set to false.
+     * The message will be `Task.upToDateWhen is false.`.
+     */
+    public static TaskExecutionMode upToDateWhenFalse() {
+        return new DefaultTaskExecutionMode("Task.upToDateWhen is false.", true, false);
+    }
+
+    /**
+     * The execution mode when the task is marked explicitly untracked.
+     * The message will be `Task state is not tracked.`.
+     */
+    public static TaskExecutionMode untracked() {
+        return new DefaultTaskExecutionMode("Task state is not tracked.", false, false);
+    }
```

Comment:
:x: We should use static constant instances instead of creating new instances wherever we can.